src/highdicom/seg/sop.py

```diff
@@ -61,6 +75,723 @@
+
+        The temporary table is named "TemporarySegmentNumbers" with columns
+        OutputSegmentNumber and SegmentNumber that are populated with values
+        derived from the input. Control flow then returns to code within the
+        "with" block. After the "with" block has completed, the cleanup of
+        the table is automatically handled.
+
+        Parameters
+        ----------
+        segment_numbers: Sequence[int]
+            Segment numbers to include, in the order desired.
+        combine_segments: bool
+            Whether the segments will be combined into a label map.
+        relabel: bool
+            Whether the output segment numbers should be relabelled to 1-n
+            (True) or retain their values in the original segmentation object.
+
+        """
+        if combine_segments:
+            if relabel:
+                # Output segment numbers are consecutive and start at 1
+                data = enumerate(segment_numbers, 1)
+            else:
+                # Output segment numbers are the same as the input
+                # segment numbers
+                data = zip(segment_numbers, segment_numbers)
+        else:
+            # Output segment numbers are indices along the output
+            # array's segment dimension, so are consecutive starting at
+            # 0
+            data = enumerate(segment_numbers)
+
+        cmd = (
+            'CREATE TABLE TemporarySegmentNumbers('
+            '    SegmentNumber INTEGER UNIQUE NOT NULL,'
+            '    OutputSegmentNumber INTEGER UNIQUE NOT NULL'
+            ')'
+        )
+
+        with self._db_con:
+            self._db_con.execute(cmd)
+            self._db_con.executemany(
+                'INSERT INTO '
+                'TemporarySegmentNumbers('
+                '    OutputSegmentNumber, SegmentNumber'
+                ')'
+                'VALUES(?, ?)',
+                data
+            )
+
+        # Yield execution to "with" block
+        yield
+
+        # Clean up table after user code executes
+        with self._db_con:
+            self._db_con.execute('DROP TABLE TemporarySegmentNumbers')
+
+    @contextmanager
+    def get_indices_by_source_instance(
+        self,
+        source_sop_instance_uids: Sequence[str],
+        segment_numbers: Sequence[int],
+        combine_segments: bool = False,
+        relabel: bool = False,
+    ) -> Generator[Iterator[Tuple[int, int, int]], None, None]:
+        """Get frame indices to create a segmentation mask for given instances.
+
+        This yields an iterator to the underlying database result, and is
+        intended to be used as a context manager for efficiency.
+
+        Parameters
+        ----------
+        source_sop_instance_uids: str
+            SOP Instance UID of the source instances to for which segmentations
+            are requested.
+        segment_numbers: Sequence[int]
+            Sequence containing segment numbers to include.
```

Comment:
I would avoid a description of the type. That information is already provided via the type declaration.

```suggestion
            Numbers of segments to include.
```